src/ranch_acceptors_sup.erl

```diff
@@ -25,28 +25,54 @@ start_link(Ref, NumAcceptors, Transport) ->
 
 init([Ref, NumAcceptors, Transport]) ->
 	TransOpts = ranch_server:get_transport_options(Ref),
-	Logger = maps:get(logger, TransOpts, error_logger),
+	NumListenSockets = maps:get(num_listen_sockets, TransOpts, 1),
 	SocketOpts = maps:get(socket_opts, TransOpts, []),
-	%% We temporarily put the logger in the process dictionary
-	%% so that it can be used from ranch:filter_options. The
-	%% interface as it currently is does not allow passing it
-	%% down otherwise.
+	Logger = maps:get(logger, TransOpts, error_logger),
+	LSockets = start_listen_sockets(Ref, NumListenSockets, Transport, SocketOpts, Logger, []),
+	[{_, BaseSocket}|_] = LSockets,
+	{ok, Addr} = Transport:sockname(BaseSocket),
+	ranch_server:set_addr(Ref, Addr),
+	Procs = lists:map(
+		fun
+			(AcceptorId) ->
+				ListenSocketId = (AcceptorId rem NumListenSockets) + 1,
+				{_, LSocket} = lists:keyfind(ListenSocketId, 1, LSockets),
+				{
+					{acceptor, self(), AcceptorId},
+					{ranch_acceptor, start_link, [Ref, AcceptorId, LSocket, Transport, Logger]},
+					permanent,
+					brutal_kill,
+					worker,
+					[]
+				}
+		end,
+		lists:seq(1, NumAcceptors)
+	),
+	{ok, {{one_for_one, 1, 5}, Procs}}.
+
+-spec start_listen_sockets(any(), pos_integer(), module(), #{}, module(), list()) ->
+	[{pos_integer(), inet:socket()}].
+start_listen_sockets(_, 0, _, _, _, Acc) ->
+	Acc;
+start_listen_sockets(Ref, N, Transport, SocketOpts, Logger, []) ->
+	LSocket = start_listen_socket(Ref, Transport, SocketOpts, Logger),
+	{ok, {_, Port}} = Transport:sockname(LSocket),
+	SocketOpts1 = [{port, Port}|lists:keydelete(port, 1, SocketOpts)],
+	start_listen_sockets(Ref, N - 1, Transport, SocketOpts1, Logger, [{N, LSocket}]);
+start_listen_sockets(Ref, N, Transport, SocketOpts, Logger, Acc) ->
+	LSocket = start_listen_socket(Ref, Transport, SocketOpts, Logger),
+	start_listen_sockets(Ref, N - 1, Transport, SocketOpts, Logger, [{N, LSocket}|Acc]).
```

Comment:
Remove port/add it to `SocketOpts` separately before calling this function and merge the two clauses. Better yet once you've done that you can just have a lists comprehension as well.